### PR TITLE
Use jinja2's StrinctUndefined to make template variables mandatory

### LINKED ISSUE #224: Environment variable in bundle aren't mandatory
Tests will fail when these environment variables aren't present but Zaza will happily allow undefined variables in bundle templates

### PR DESCRIPTION
**Pull request by ChrisMacNaughton**
Closes #224

### CODE REVIEW

**Comment by ChrisMacNaughton** [comment]
I'm working on tests for this now but it should make template variables mandatory once it's done

File: zaza/charm_lifecycle/deploy.py [hunk 1]
```diff
@@ -118,7 +118,8 @@ def get_jinja2_env():
     """
     template_dir = get_overlay_template_dir()
     return jinja2.Environment(
-        loader=jinja2.FileSystemLoader(template_dir)
+        loader=jinja2.FileSystemLoader(template_dir),
+        undefined=jinja2.StrictUndefined
```
Comment (ajkavanagh) [inline zaza/charm_lifecycle/deploy.py]: Does this have the potential to break assumptions? i.e. some tests don't bother setting things as it's not needed for that scenario?

Comment (ChrisMacNaughton) [reply zaza/charm_lifecycle/deploy.py]: It shouldn't, as we only use this to render jinja variables into bundles, and those are per test anyway?

Comment (ChrisMacNaughton) [reply zaza/charm_lifecycle/deploy.py]: definitely open for discussion, but I'm not entirely certain of other ways to ensure that we require variables that are set in the test bundle's overlay

Comment (ajkavanagh) [reply zaza/charm_lifecycle/deploy.py]: Okay, I'm largely convinced.  Therefore, this is +1 from me.

File: zaza/charm_lifecycle/deploy.py [hunk 2]
```diff
@@ -158,10 +159,14 @@ def render_template(template, target_file):
...
-    with open(target_file, "w") as fh:
-        fh.write(
-            template.render(get_template_overlay_context()))
-
+    try:
+        with open(target_file, "w") as fh:
+            fh.write(
+                template.render(get_template_overlay_context()))
+    except jinja2.exceptions.UndefinedError as e:
+        logging.error("Template error. You may be mmissing"
```
Comment (thedac) [inline zaza/charm_lifecycle/deploy.py]: s/mmissing/missing

File: unit_tests/test_zaza_charm_lifecycle_deploy.py [hunk 3]
```diff
@@ -128,6 +128,23 @@ def test_render_overlay(self):
...
+        self.patch_object(lc_deploy.sys, 'exit')
+        self.patch_object(lc_deploy.logging, 'error')
+        jinja2_env = lc_deploy.get_jinja2_env()
+        template = jinja2_env.from_string('{{required_variable}}')
+        m = mock.mock_open()
+        with mock.patch('zaza.charm_lifecycle.deploy.open', m, create=True):
+            lc_deploy.render_template(template, '/tmp/mybundle.yaml')
+        m.assert_called_once_with('/tmp/mybundle.yaml', 'w')
+        self.error.assert_called_once_with(
+            "Template error. You may be mmissing"
```
Comment (thedac) [inline unit_tests/test_zaza_charm_lifecycle_deploy.py]: s/mmissing/missing

**Comment by davecore82** [comment]
I tested this PR with the keystone change without setting AMULET_OS_VIP00 and it works as expected:

> $ tox -e func --workdir /tmp
> [...]
> INFO:root:Deploying bundle './tests/bundles/trusty-mitaka.yaml' on to 'zaza-946d3f1a24bc' model
> ERROR:root:Template error. You may be mmissinga mandatory environment variable : 'AMULET_OS_VIP00' is undefined
> ERROR: InvocationError: '/tmp/func/bin/functest-run-suite --keep-model'
> [...]
> ERROR:   func: commands failed

When I set the AMULET_OS_VIP00 var, the test proceeds normally. Thanks!

File: zaza/charm_lifecycle/deploy.py (hunk 2 above)
Comment (ChrisMacNaughton) [reply zaza/charm_lifecycle/deploy.py]: This is resolved in the latest patch

**Pull request merged**